### ISSUE TITLE
linter: require-array-sort-compare doesn't detect conditional generic string types

### ISSUE THREAD

**Issue by zachkirsch** [oxc-project/oxc#26566]
### What version of Oxlint are you using?

1.82.0

### What command did you run?

oxlint --type-aware --deny-warnings --report-unused-disable-directives --ignore-pattern 'packages/**'

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

```jsonc
const config = defineConfig({
  plugins: ["typescript", "unicorn", "import", "oxc", "promise"],
  ...
});
```

### What happened?

noticed that oxlint will flag false positives for conditional string types - i..e. conditional types where both branches are strings. Minimal example:

```ts
function sort<T>(values: (T extends 0 ? string : string)[]) {
  values.sort();
}
```

warning:

```
server:turbo:lint:   ⚠ typescript(require-array-sort-compare): Require 'compare' argument.
server:turbo:lint:     ╭─[<filename>:77:3]
server:turbo:lint:  76 │ function sort<T>(values: (T extends 0 ? string : string)[]) {
server:turbo:lint:  77 │   values.sort();
server:turbo:lint:     ·   ─────────────
server:turbo:lint:  78 │ }
server:turbo:lint:     ╰────
server:turbo:lint: 
server:turbo:lint: Found 1 warning and 2 errors.
server:turbo:lint: Finished in 13.1s on 1541 files with 235 rules using 32 threads.
```

**Comment by im10furry** [comment]
I checked this against upstream `typescript-eslint` with the rule enabled explicitly (`parserOptions.projectService`, both on the same file):

| input | oxlint `--type-aware` | typescript-eslint |
| --- | --- | --- |
| `(T extends 0 ? string : string)[]` | reports | **reports** |
| `string[]` | – | – |
| `("x" \| "y")[]` | – | – |
| `T[]` | reports | reports |
| `T extends string`, `T[]` | – | – |
| `(T extends string ? T : never)[]` | reports | reports |
| `String[]` | reports | reports |

So this is not an oxlint/tsgolint divergence: the conditional type's flags are `Conditional` (not `StringLike`), so `ignoreStringArrays` does not apply in either implementation, and upstream reports it too. (`getBaseConstraintOfType` resolves it to `string`, but neither implementation consults that.) Relaxing it here would be a behaviour change relative to upstream, so it probably needs to go to typescript-eslint first.

Separately — while checking parity around this rule I did hit one **real** divergence, unrelated to conditional types: bare tuple types are reported by oxlint but not by upstream.

| input | oxlint | typescript-eslint |
| --- | --- | --- |
| `[string, number]`, `[number, string]`, `readonly [string, number]` | reports | – |
| `[string, ...number[]]`, `[string, number?]` | reports | – |
| `[string, number] \| number[]` | reports | – |
| `Array<[string, number]>` | reports | reports |

Upstream's report condition is `isTypeArrayTypeOrUnionOfArrayTypes` = every union constituent must satisfy `checker.isArrayType()`, which excludes tuples; tsgolint's final check uses `isArrayOrTupleType`. Neither repo's test suite covers tuples, which is probably why it went unnoticed. That one belongs in oxc-project/tsgolint rather than here.

**Comment by im10furry** [comment]
Follow-up: the tuple divergence I mentioned is now opened as oxc-project/tsgolint#1180.

**Comment by camc314** [comment]
this should be fixed - please checkout the latest oxlint-tsgolint release > https://x.com/Cameron_C2/status/2100831510526083417

**Comment by zachkirsch** [comment]
Thank you @camc314 !!

**Comment by zachkirsch** [comment]
@camc314 unfortunately this is still broken in v7.0.2002

```ts
function sort<T>(values: (T extends 0 ? string : string)[]) {
  values.sort();
}
```

**Comment by im10furry** [comment]
Re-checked on `oxlint 1.83.0` + `oxlint-tsgolint 7.0.2002`: it does still reproduce, but it turns out this isn't an oxlint/tsgolint divergence — upstream `@typescript-eslint/require-array-sort-compare` reports the exact same line (8.70.0, `projectService`, `ignoreStringArrays: true`), and the two agree line-for-line across the whole matrix: the conditional variants, bare tuples, `Array<[string, number]>`, `("x" | "y")[]`, `T extends string; T[]`, `T[]`.

7.0.2002 only contained #1180 (`isArrayOrTupleType` → `isArrayType`) — that was the tuple divergence from my earlier comment, not this one. Nothing in the rule's string detection changed in that release.

Root cause on both sides: the element type is a deferred `Conditional` type (`TypeFlags` 16777216) and `getTypeName` has no conditional branch, so `ignoreStringArrays` can never apply. Neither implementation consults `getBaseConstraintOfType`, which resolves this element to `string`.

So this needs to be fixed upstream first: typescript-eslint/typescript-eslint#12909. I'm happy to port it here once it lands.

**Comment by zachkirsch** [comment]
ah bummer, thanks for the intel!
